packages/uploadthing/test/__test-helpers.ts

```diff
@@ -94,111 +98,269 @@ const mockPresigned = (file: {
+      headers: Object.fromEntries(request.headers.entries()),
     });
+
+    return HttpResponse.json({ success: true });
   }),
 ];
 
-const s3Api = [
-  http.post("https://bucket.s3.amazonaws.com/", ({ params, request }) => {
-    s3Mock(params, request);
-    return HttpResponse.json(null, { status: 204 });
+const staticAssetServer = [
+  http.get("https://cdn.foo.com/:fileKey", ({ request }) => {
+    staticAssetMock(request.url);
+
+    return HttpResponse.text("Lorem ipsum doler sit amet", {
+      headers: { "content-type": "text/plain" },
+    });
   }),
-  http.put("https://bucket.s3.amazonaws.com/:key", ({ params, request }) => {
-    s3Mock(params, request);
-    return HttpResponse.json(null, {
-      status: 204,
-      headers: { ETag: "abc123" },
+  http.get("https://utfs.io/f/:key", ({ request }) => {
+    staticAssetMock(request.url);
+
+    return HttpResponse.text("Lorem ipsum doler sit amet", {
+      headers: { "content-type": "text/plain" },
     });
   }),
 ];
 
-export const handlers = [...staticAssetServer, ...s3Api];
+const s3Mocks = (fail: boolean) =>
+  fail
+    ? [
+        http.post("https://bucket.s3.amazonaws.com", async ({ request }) => {
+          s3Mock(request.url, {
+            method: request.method,
+            headers: Object.fromEntries(request.headers.entries()),
+            body: await request.formData(),
+          });
+          return HttpResponse.json(null, { status: 403 });
+        }),
+        http.put(
+          "https://bucket.s3.amazonaws.com/:key",
+          async ({ request }) => {
+            s3Mock(request.url, {
+              method: request.method,
+              headers: Object.fromEntries(request.headers.entries()),
+              body: await request.formData(),
+            });
+            return HttpResponse.json(null, { status: 204 });
+          },
+        ),
+      ]
+    : [
+        http.post("https://bucket.s3.amazonaws.com", async ({ request }) => {
+          s3Mock(request.url, {
+            method: request.method,
+            headers: Object.fromEntries(request.headers.entries()),
+            // body: fd,
+          });
+          return HttpResponse.json(null, { status: 204 });
+        }),
+        http.put(
+          "https://bucket.s3.amazonaws.com/:key",
+          async ({ request }) => {
+            s3Mock(request.url, {
+              method: request.method,
+              headers: Object.fromEntries(request.headers.entries()),
+              body: await request.formData(),
+            });
+            return HttpResponse.json(null, {
+              status: 204,
+              headers: { ETag: "abc123" },
+            });
+          },
+        ),
+      ];
+
+export const setupMsw = (opts: {
+  db: MockDbInterface;
+  failS3Call?: boolean;
+}) => {
+  const server = setupServer(
+    ...s3Mocks(opts.failS3Call ?? false),
+    ...staticAssetServer,
+    ...utapiMocks(opts.db),
+  );
+  server.listen({ onUnhandledRequest: "warn" });
```

Comment:
If you plan on running tests concurrently, I highly recommend using the [`server.boundary()` API](https://mswjs.io/docs/api/setup-server/boundary) to scope each `server.use()` modifications to the individual test's scope.

Reply:
good to know, thanks!